### PR TITLE
feat: Report notification metrics for rollouts

### PR DESCRIPTION
**Pull request by RaviHari**
Signed-off-by: Ravi Hari <ravireliable@gmail.com>
Fixes: https://github.com/argoproj/argo-rollouts/issues/1782

Notification_Metrics

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [X] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [X] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [X] My builds are green. Try syncing with master if they are not. 
* [X] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

File: controller/metrics/prommetrics.go [hunk 1]
```diff
@@ -175,6 +175,34 @@ var (
...
+		prometheus.CounterOpts{
+			Name: "notification_send_error",
+			Help: "Error sending the notification",
+		},
+		append(namespaceNameLabels, "type", "reason"),
+	)
+
+	MetricNotificationSendPerformance = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "notification_send",
```
Comment (agrawroh) [inline controller/metrics/prommetrics.go:198]: You want to call it `notification_send_performance` or `notification_send`?

**Review by agrawroh — COMMENTED** [review]
Would you please add the new stats description to the Doc as well?

**Review by agrawroh — COMMENTED** [review]
Looks good. Thanks for adding the docs. Would you please also add some tests to verify that new counters are working as expected?

File: utils/record/record_test.go [hunk 2]
```diff
@@ -100,11 +101,87 @@ func TestSendNotifications(t *testing.T) {
...
+	want := `# HELP notification_send_performance Notification send performance.
+# TYPE notification_send_performance histogram
+notification_send_performance_bucket{name="guestbook",namespace="default",le="0.01"} 0
+notification_send_performance_bucket{name="guestbook",namespace="default",le="0.15"} 1
+notification_send_performance_bucket{name="guestbook",namespace="default",le="0.25"} 3
+notification_send_performance_bucket{name="guestbook",namespace="default",le="0.5"} 6
+notification_send_performance_bucket{name="guestbook",namespace="default",le="1"} 8
+notification_send_performance_bucket{name="guestbook",namespace="default",le="+Inf"} 11
+notification_send_performance_sum{name="guestbook",namespace="default"} 7.27
+notification_send_performance_count{name="guestbook",namespace="default"} 11
```
Comment (agrawroh) [inline utils/record/record_test.go:179]: Please indent this block.

File: utils/record/record.go [hunk 3]
```diff
@@ -207,6 +239,13 @@ func NewAPIFactorySettings() api.Settings {
 // Send notifications for triggered event if user is subscribed
 func (e *EventRecorderAdapter) sendNotifications(object runtime.Object, opts EventOptions) error {
 	logCtx := logutil.WithObject(object)
+	_, namespace, name := logutil.KindNamespaceName(logCtx)
+	startTime := timeutil.Now()
+	defer func() {
+		duration := time.Since(startTime)
+		e.NotificationSendPerformance.WithLabelValues(namespace, name).Observe(duration.Seconds())
+		logCtx.WithField("time_ms", duration.Seconds()*1e3).Info("Notification sent")
```
Comment (agrawroh) [inline utils/record/record.go:247]: `Info` might be too verbose. Could you please change it to `Debug`?

**Review by agrawroh — APPROVED** [review]
LGTM. Thank you so much for working on this!

File: utils/record/record.go (hunk 3 above)
Comment (RaviHari) [reply utils/record/record.go]: Thanks fixed the comments.

**Review by harikrongali — APPROVED** [review]

**Comment by harikrongali** [comment]
@alexmt can you please merge the PR

**Review by alexmt — APPROVED** [review]

**Pull request merged**